### PR TITLE
Add macOS compatibility

### PR DESCRIPTION
**Pull request by ygoumaz**
Currently, `webots_ros2` cannot be used on macOS. This PR aims to add the necessary framework to allow such compatibility.

The solution is as follows:
- The user needs to install Docker on its machine.
- A Docker container based on the ROS2 image (arm64) is created.
- The Docker container shares a folder with the host and contains a special environment variable (`WEBOTS_SHARED_FOLDER`) that allows `webots_ros2` to detect the macOS environment.
- `webots_ros2` must be installed in the container.
- In this case (unlike Linux and Windows), Webots is not launched directly from the package but using the shared folder. A Webots "server" is used to start Webots, with the world and the additional arguments found in the shared volume.
- The connection of the controller(s) is possible through TCP.

### CODE REVIEW

**Comment by ygoumaz** [comment]
The PR is ready for review. In the meantime, I will add the documentation to the wiki. The webots server to start the simulation on the host is also ready for review here: https://github.com/cyberbotics/webots-server/pull/13.

File: webots_ros2_driver/webots_ros2_driver/webots_launcher.py
Comment (omichel) [inline webots_ros2_driver/webots_ros2_driver/webots_launcher.py:198]:
```suggestion
            tcp_url = 'host.docker.internal'
```

File: webots_ros2_universal_robot/launch/robot_launch/robot_nodes_launch.py
Comment (omichel) [inline webots_ros2_universal_robot/launch/robot_launch/robot_nodes_launch.py:39]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_universal_robot/launch/multirobot_launch.py
Comment (omichel) [inline webots_ros2_universal_robot/launch/multirobot_launch.py:47]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_turtlebot/launch/robot_launch.py
Comment (omichel) [inline webots_ros2_turtlebot/launch/robot_launch.py:44]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_tiago/launch/robot_launch.py
Comment (omichel) [inline webots_ros2_tiago/launch/robot_launch.py:53]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_tests/test/test_system_driver.py
Comment (omichel) [inline webots_ros2_tests/test/test_system_driver.py:55]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_tesla/launch/robot_launch.py
Comment (omichel) [inline webots_ros2_tesla/launch/robot_launch.py:42]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_mavic/launch/robot_launch.py
Comment (omichel) [inline webots_ros2_mavic/launch/robot_launch.py:42]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_epuck/launch/robot_launch.py
Comment (omichel) [inline webots_ros2_epuck/launch/robot_launch.py:44]:
```suggestion
        tcp_url = 'host.docker.internal'
```

File: webots_ros2_driver/webots_ros2_driver/utils.py
Comment (omichel) [inline webots_ros2_driver/webots_ros2_driver/utils.py:110]: I would replace this with:
```suggestion
def has_shared_folder():
```
as it may not be always specific to macOS.

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by ygoumaz** [comment]
The [wiki page](https://github.com/cyberbotics/webots_ros2/wiki/Complete-Installation-Guide) is now updated with the instructions to use `webots_ros2` on macOS.

**Review by omichel — CHANGES_REQUESTED** [review]
I would rename `WEBOTS_SHARED_FOLDERS` to `WEBOTS_SHARED_FOLDER`.
Also, the `FAIL:x` error messaging could be improved: instead of `FAIL:0` we could send `FAIL: Error in clear text`.
Finally, the local server could send notifications when Webots quits, etc. so that it behaves the same as on WSL and Linux.

**Comment by ygoumaz** [comment]
The PR is finally ready for review again. I refactored quite a lot the code:

- The connection to the Webots starting server is now done in a python subprocess (`webots_tcp_client.py`). This way, when Webots is closed the subprocess receives a message and exits. The other nodes can be shutdown too.
- The shared folder is now cleaned from the ROS side and not from the server side.
- Failure messages are sent through TCP.
- In the universal robot example, the reset handler (which respawns all nodes when the simulation is reset) must be unregistered when closing webots to avoid additional copy of resource files to the shared folder.
- Bug fixes and cleanups.

**Comment by ygoumaz** [comment]
I will now improve the documentation by providing a Dockerfile to install the complete environment.

**Review by omichel — APPROVED** [review]
It seems good to me.

I only have a doubt regarding the way the arguments are passed to the local simulation server.
Currently they are stored in a file in the shared folder with a specific name (and a small risk of name clash).

I was wondering whether it could be better (more generic) to send them to the local simulation server as the first connection message, e.g., the message would mean something like "please start webots with the following command line parameters including the world filename", it could take the following form: `COMMAND:webots --batch --mode=RUN /my/shared/folder/worlds/my_world.wbt`. This would simplify the checks on the server side and would be more generic, e.g., allow to launch any command on the host machine. For security reasons we should however, check that the command, e.g. "webots" is in a list of accepted commands.

**Comment by ygoumaz** [comment]
I will apply these changes. You are right that is will be a more generic and clean solution. Meanwhile, the wiki page is now updated.  The Docker installation is now summarized to a Dockerfile to build.

File: webots_ros2_driver/scripts/webots_tcp_client.py
Comment (omichel) [inline webots_ros2_driver/scripts/webots_tcp_client.py:40]: So we are not passing the full path to the webots binary here?
This assume the local server script should prepend the WEBOTS_HOME to it?
Do we want this?

Comment (ygoumaz) [reply webots_ros2_driver/scripts/webots_tcp_client.py]: As discussed, here we only send 'webots' as executable. The server is responsible to append the WEBOTS_HOME prefix.

**Review by omichel — APPROVED** [review]
Nice.

**Comment by ygoumaz** [comment]
The tests are failing, because the last nightly of Webots is not available. We can wait until Monday to see if all tests pass to merge the PR.

**Pull request merged**
